Telemetry/TelemetryScheduler.swift

```diff
@@ -26,28 +26,32 @@ public class TelemetryScheduler {
             return
         }
 
-        func sequentialUploader() {
+        func sequentialPingUploader() {
             guard !hasReachedDailyUploadLimitForPingType(pingType),
                 pings.count > 0,
                 let ping = TelemetryPing.from(dictionary: pings.remove(at: 0)) else {
                 completionHandler()
                 return
             }
 
-            client.upload(ping: ping) { error in
-                if let error = error {
-                    print("Error uploading TelemetryPing: \(error.localizedDescription)")
+            client.upload(ping: ping) { httpStatusCode, error in
+                let errorCode = (error as NSError?)?.code ?? 0
+                let errorRequiresDelete = [TelemetryError.InvalidUploadURL, TelemetryError.CannotGenerateJSON].contains(errorCode)
+
+                // Arguably, this could be (200..<500).contains(httpStatusCode) and 5xx errors could be handled more selectively to decide whether to delete the ping.
+                if httpStatusCode >= 200 || errorRequiresDelete {
```

Comment:
This doesn't seem right? We're still potentially throwing away all pings here regardless of this conditional. We're just being selective about when we write to storage. In the code above on line 32, we still remove the ping from the array, so if `sequentialPingUploader` gets called multiple times, a `500` upload followed by a `200` upload would still end up deleting the first one. This is fine for right now anyway because this is more of a problem in `TelemetryStorage` than anything else. We'll get this right once we move to one-file-per-ping which I'm in the middle of doing.

Reply:
> potentially throwing away all pings here regardless of this conditional

I hope not, the intention is that a ping is removed from disk only only as a result of conditional on line 42. 

> In the code above on line 32, we still remove the ping from the array, so if sequentialPingUploader gets called multiple times

Line 49, the sequentialPingUploader loop breaks out. That is, in the event of an error that does not match conditional on line 42, the entire upload process is stopped. The only possible errors at this point are intermittent network failure (or coding error). The next time the app schedules an upload (currently when backgrounded) it will try again, with the same data.

Reply:
Oh right. I forgot this breaks the loop on an error. That's also unfortunate, but again we can fix this with file-based ping storage.

Reply:
agreed, I'll get to that in the next day or two

Reply:
i already started on a patch for this

Reply:
i can probably post a PR tomorrow.. thanks for opening the issue.. i'll assign myself so we can keep track